icu4c/source/i18n/coll.cpp

```diff
@@ -746,7 +746,7 @@ UObject*
 CFactory::create(const ICUServiceKey& key, const ICUService* /* service */, UErrorCode& status) const
 {
     if (handlesKey(key, status)) {
-        const LocaleKey& lkey = (const LocaleKey&)key;
+        const LocaleKey& lkey = dynamic_cast<const LocaleKey&>(key);
```

Comment:
If I change this line to static_cast<> then I got the following runtime error during test under CFI
```
      CollationServiceTest {
         TestRegister {
         } OK:   TestRegister 
         TestRegisterFactory {
coll.cpp:749:33: runtime error: control flow integrity check for type 'icu::LocaleKey' failed during base-to-derived cast (vtable address 0x7fb135bfa730)
0x7fb135bfa730: note: invalid vtable
 b1 7f 00 00  a0 7e ac 35 b1 7f 00 00  80 7b ac 35 b1 7f 00 00  a0 7b ac 35 b1 7f 00 00  f0 59 ac 35
              ^ 
coll.cpp:749:33: note: check failed in ../../lib/libicui18n.so.73, vtable located in ../../lib/libicuuc.so.73
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior coll.cpp:749:33 in 

```